components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+                    actionInvocationResponse.getErrorLog() != null ? actionInvocationResponse.getErrorLog() :
+                            "Unknown"));
+        }
+    }
+
+    private String serializeRequest(ActionExecutionRequest request) throws JsonProcessingException {
+
+        ObjectMapper requestObjectmapper = new ObjectMapper();
+        requestObjectmapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        requestObjectmapper.setSerializationInclusion(JsonInclude.Include.NON_EMPTY);
+        return requestObjectmapper.writeValueAsString(request);
+    }
+
+    private String serializeSuccessResponse(ActionInvocationSuccessResponse response) throws JsonProcessingException {
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        return objectMapper.writeValueAsString(response);
+    }
+
+    private String serializeErrorResponse(ActionInvocationErrorResponse response) throws JsonProcessingException {
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        return objectMapper.writeValueAsString(response);
+    }
+
+    private List<PerformableOperation> validatePerformableOperations(ActionExecutionRequest request,
+                                                                     ActionInvocationSuccessResponse response) {
+
+        List<AllowedOperation> allowedOperations = request.getAllowedOperations();
+
+        List<PerformableOperation> allowedPerformableOperations = response.getOperations().stream()
+                .filter(performableOperation -> allowedOperations.stream()
+                        .anyMatch(allowedOperation -> OperationComparator.compare(allowedOperation,
+                                performableOperation)))
+                .collect(Collectors.toList());
+
+        if (LOG.isDebugEnabled()) {
+            // todo: add to diagnostics
+            List<String> allowedOps = new ArrayList<>();
+            List<String> notAllowedOps = new ArrayList<>();
+
+            response.getOperations().forEach(operation -> {
+                String operationDetails = "Operation: " + operation.getOp() + " with path: " + operation.getPath();
+                if (allowedPerformableOperations.contains(operation)) {
+                    allowedOps.add(operationDetails);
+                } else {
+                    notAllowedOps.add(operationDetails);
+                }
+            });
+
+            String logMessage = "Allowed Operations: " + String.join(", ", allowedOps) +
+                    ". Not Allowed Operations: " + String.join(", ", notAllowedOps);
+            LOG.debug(logMessage);
+        }
+
+        return allowedPerformableOperations;
+    }
+
+    private AuthMethods.AuthMethod getAuthenticationMethod(String actionId, AuthType authType)
+            throws ActionMgtException {
+
+        List<AuthProperty> authProperties = authType.getPropertiesWithDecryptedValues(actionId);
+
+        switch (authType.getType()) {
+            case BASIC:
```

Comment:
Where it defines these constants ?
BASIC
BEARER
API_KEY
NONE

Reply:
Does the NONE capture null ? or empty ?

Reply:
These auth types are defined in the action management component. Only application of the authentication applies here.

NONE means authentication is not needed

Reply:
So we just call the endpoint